plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsReader.java

```diff
@@ -85,33 +86,38 @@
 
     public static final String APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY = "ndv";
 
-    private final TypeManager typeManager;
-    private final ConnectorSession session;
-    private final Table icebergTable;
-
-    private TableStatisticsReader(TypeManager typeManager, ConnectorSession session, Table icebergTable)
-    {
-        this.typeManager = typeManager;
-        this.session = session;
-        this.icebergTable = icebergTable;
-    }
+    private TableStatisticsReader()
+    { }
 
     public static TableStatistics getTableStatistics(TypeManager typeManager, ConnectorSession session, IcebergTableHandle tableHandle, Table icebergTable)
     {
-        return new TableStatisticsReader(typeManager, session, icebergTable).makeTableStatistics(tableHandle);
+        return makeTableStatistics(
+                typeManager,
+                icebergTable,
+                tableHandle.getSnapshotId(),
+                tableHandle.getEnforcedPredicate(),
+                tableHandle.getUnenforcedPredicate(),
+                isExtendedStatisticsEnabled(session));
     }
 
-    private TableStatistics makeTableStatistics(IcebergTableHandle tableHandle)
+    @VisibleForTesting
+    public static TableStatistics makeTableStatistics(
+            TypeManager typeManager,
+            Table icebergTable,
+            Optional<Long> snapshot,
+            TupleDomain<IcebergColumnHandle> enforcedConstraint,
+            TupleDomain<IcebergColumnHandle> unenforcedConstraint,
+            boolean extendedStatisticsEnabled)
     {
-        if (tableHandle.getSnapshotId().isEmpty()) {
+        if (snapshot.isEmpty()) {
             // No snapshot, so no data.
             return TableStatistics.builder()
                     .setRowCount(Estimate.of(0))
                     .build();
         }
-        long snapshotId = tableHandle.getSnapshotId().get();
+        long snapshotId = snapshot.get();
 
-        TupleDomain<IcebergColumnHandle> enforcedPredicate = tableHandle.getEnforcedPredicate();
+        TupleDomain<IcebergColumnHandle> enforcedPredicate = enforcedConstraint.intersect(unenforcedConstraint);
```

Comment:
I'm not a 100% sure, but I think this might lead to underestimation by the CBO.
Even though `unenforcedConstraint` was used to prune statistics here, that filter is going to stay on top of the scan for evaluation. FilterStatsCalculator will estimate it like any other predicate and further reduce the already filtered stats. Please check if this is the case.
cc: @findepi @sopel39

Reply:
@raunaqmorarka the code here seems correct to me.
Note that the logic here should be -- and is -- aligned with how we select files for reading later on 
https://github.com/trinodb/trino/blob/75bc783419482f30c103cf74510498ff480e82fc/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java#L167-L178

Reply:
I get that the logic here matches with file pruning in splits generation. My concern was about something else.
E.g.: Let's say before this change FilterStatsCalculator would get column stats based on 10 file's stats and uses that to estimate a predicate. After this change FilterStatsCalculator will estimate the same predicate but with column stats based on a subset of file's stats. The domain of the column stats (min/max, ndv) would be narrower this time.
This is still correct though as the row count will also be smaller and we assume uniform distribution of values.
It does change the estimates when actual distribution of values across files is non-uniform, maybe this is why a couple of TPC queries plans also changed.

Reply:
I see your concern. Yes, this may result in different stats eventually calculated by FilterStatsCalculator.
Yet, this is IMO the correct thing to do. If the connector is going to return rows from 10 files, it should indicate row count coming from 100s of files.